cloud-controller-manager/do/firewall_controller.go

```diff
@@ -0,0 +1,365 @@
+			if err != nil {
+				return fmt.Errorf("could not create firewall: %v", err)
+			}
+			klog.Info("successfully created firewall")
+		}
+		fm.fwCache.updateCache(currentFirewall)
+		return nil
+	}
+
+	// Check if the target firewall ID exists but in the case that CCM first starts up and the
+	// firewall ID does not exist yet, we have to retrieve one from the API.
+	if targetFirewall == nil {
+		currentFirewall, err := fm.Get(ctx)
+		if err != nil {
+			return err
+		}
+		if currentFirewall == nil {
+			klog.Info("an existing firewall not found, we need to create one")
+			currentFirewall, err = fm.createFirewall(ctx, fr.InboundRules)
+			if err != nil {
+				return err
+			}
+			klog.Info("successfully created firewall")
+		}
+		fm.fwCache.updateCache(currentFirewall)
+	}
+	return nil
+}
+
+func isEqual(targetFirewall *godo.Firewall, fr *godo.FirewallRequest) bool {
+	equal := true
+	if !cmp.Equal(targetFirewall.Name, fr.Name) {
+		klog.Info("reconciling firewall name")
+		equal = false
+	} else if !cmp.Equal(targetFirewall.InboundRules, fr.InboundRules) {
+		klog.Info("reconciling inbound rules")
+		equal = false
+	} else if !cmp.Equal(targetFirewall.OutboundRules, fr.OutboundRules) {
+		klog.Info("reconciling outbound rules")
+		equal = false
+	} else if !cmp.Equal(targetFirewall.Tags, fr.Tags) {
+		klog.Info("reconciling firewall tags")
+		equal = false
+	}
+	return equal
+}
+
+func (fm *firewallManagerOp) createFirewall(ctx context.Context, svcInboundRules []godo.InboundRule) (*godo.Firewall, error) {
+	fr := createFirewallRequest(svcInboundRules, fm.workerFirewallName, fm.workerFirewallTags)
+	currentFirewall, _, err := fm.client.Firewalls.Create(ctx, fr)
+	return currentFirewall, err
+}
+
+// create a firewall request that has the correct rules, name and tag
+func (fc *FirewallController) createReconciledFirewallRequest(serviceList []*v1.Service) *godo.FirewallRequest {
+	var nodePortInboundRules []godo.InboundRule
+	for _, svc := range serviceList {
+		if svc.Spec.Type == v1.ServiceTypeNodePort {
+			// this is a nodeport service so we should check for existing inbound rules on all ports.
+			for _, servicePort := range svc.Spec.Ports {
+				// In the odd case that a failure is asynchronous causing the NodePort to be set to zero.
+				if servicePort.NodePort == 0 {
+					klog.Warningf("NodePort on the service is set to zero")
+					continue
+				}
+				nodePortInboundRules = append(nodePortInboundRules,
+					godo.InboundRule{
+						Protocol:  "tcp",
+						PortRange: strconv.Itoa(int(servicePort.NodePort)),
+						Sources: &godo.Sources{
+							Tags: fc.workerFirewallTags,
+						},
+					},
+					godo.InboundRule{
+						Protocol:  "udp",
+						PortRange: strconv.Itoa(int(servicePort.NodePort)),
+						Sources: &godo.Sources{
+							Tags: fc.workerFirewallTags,
+						},
+					},
+				)
+			}
+		}
+	}
+	return createFirewallRequest(nodePortInboundRules, fc.workerFirewallName, fc.workerFirewallTags)
```

Comment:
I don't think we need to pass in `fc.workerFirewallName` and `fc.workerFirewallTags` if we turn `createFirewallRequest` into a method on `* FirewallController` and reference the variables from the receiver.

Reply:
We also call it here, which does not have access to the FirewallController:
```go
func (fm *firewallManagerOp) createFirewall(ctx context.Context, svcInboundRules []godo.InboundRule) (*godo.Firewall, error) {
	fr := createFirewallRequest(svcInboundRules, fm.workerFirewallName, fm.workerFirewallTags)
	currentFirewall, _, err := fm.client.Firewalls.Create(ctx, fr)
	return currentFirewall, err
}
```

and the above func should keep the same receiver because we need `fm.client.Firewalls.Create(ctx, fr)` @timoreimann

Reply:
After the latest refactorings, it looks like this is the only spot where we invoke `createFirewallRequest`. Could we now turn it into a method and reference the firewall name and tag from the receiver?